dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -260,71 +209,115 @@ private void simplifyAssociatives(Class<? extends Definition> cls, BiFunction<Re
+        @Override
+        public Void visitConstraint(Constraint constraint) {
+            for (Relation rel :  constraint.getConstrainedRelations()) {
+                if (collectedRelations.add(rel)) {
+                    rel.getDefinition().accept(this);
+                }
+            }
+            return null;
+        }
+    }
+
+    private Relation makePredefinedRelation(String name) {
+        /*
+            WARNING: The code has possibly unexpected behaviour:
+             - If a relation like `ext` is already defined (different from the standard definition),
+               constructing e.g. `rfe` will refer to this new definition.
+             - When constructing derived relations like fr, the intermediately defined relations like rf^-1
+               will get constructed even if they already exist in the model.
+             TODO: Clarify what the intended behaviour should be.
+         */
+        final Relation r = newRelation(name);
+        final Definition def = switch (name) {
             case PO -> new ProgramOrder(r, Filter.byTag(Tag.VISIBLE));
-            case LOC -> new SameAddress(r);
-            case ID -> new Identity(r, Filter.byTag(Tag.VISIBLE));
-            case INT -> new SameThread(r);
-            case EXT -> new DifferentThreads(r);
+            case LOC -> new SameLocation(r);
+            case ID -> new SetIdentity(r, Filter.byTag(Tag.VISIBLE));
+            case INT -> new Internal(r);
+            case EXT -> new External(r);
             case CO -> new Coherence(r);
             case RF -> new ReadFrom(r);
             case RMW -> new ReadModifyWrites(r);
-            case CASDEP -> new CompareAndSwapDependency(r);
-            case CRIT -> new CriticalSections(r);
+            case CASDEP -> new CASDependency(r);
+            case CRIT -> new LinuxCriticalSections(r);
             case IDD -> new DirectDataDependency(r);
             case ADDRDIRECT -> new DirectAddressDependency(r);
             case CTRLDIRECT -> new DirectControlDependency(r);
             case EMPTY -> new Empty(r);
             case RFINV -> {
                 //FIXME: We don't need a name for "rf^-1", this is a normal relation!
                 // This causes models that explicitly mention "rf^-1" to have two versions of the same relation!
-                yield new Inverse(r, getRelation(RF));
+                yield new Inverse(r, getOrCreatePredefinedRelation(RF));
             }
-            case FR -> composition(r, getRelation(RFINV), getRelation(CO));
-            case MM -> new CartesianProduct(r, Filter.byTag(Tag.MEMORY), Filter.byTag(Tag.MEMORY));
-            case MV -> new CartesianProduct(r, Filter.byTag(Tag.MEMORY), Filter.byTag(Tag.VISIBLE));
-            case IDDTRANS -> new TransitiveClosure(r, getRelation(IDD));
-            case DATA -> intersection(r, getRelation(IDDTRANS), getRelation(MM));
+            case FR -> composition(r, getOrCreatePredefinedRelation(RFINV), getOrCreatePredefinedRelation(CO));
```

Comment:
What is the difference between `getOrCreatePredefinedRelation` and `getOrCreateRelation`?

Reply:
The latter doesn't exist. But if it did exist, I would interpret it as "get relation with name X and if it does not exist, create a relation with name X". It would NOT automatically associate any definition to that relation and so would NOT implicitly create any other relations.

Reply:
Mmmm from the top of my head, I was sure I saw that method while I was reviewing this PR, but you are right.